tensorflow/compiler/xla/service/gpu/matmul_utils.cc

```diff
@@ -752,53 +804,112 @@ Status MatmulPlan::DoMatmul(se::Stream* stream, se::DeviceMemoryBase a_buffer,
-                                 bias_buffer, algorithm, scratch_allocator,
-                                 profile_result);
-    case CUDA_C_64F:
-      return DoMatmul<complex128>(stream, a_buffer, b_buffer, c_buffer,
-                                  d_buffer, bias_buffer, algorithm,
-                                  scratch_allocator, profile_result);
-    default:
-      return InternalError("Unexpected dtype");
+  std::tuple<cudaDataType_t, cudaDataType_t, cudaDataType_t, cudaDataType_t>
+      operand_types{plan_.a_desc.type(), plan_.b_desc.type(),
+                    plan_.c_desc.type(), plan_.d_desc.type()};
+
+#define TYPED_MATMUL_F8(ATYPE, BTYPE, CTYPE, DTYPE)                         \
+  if (operand_types == std::make_tuple(ATYPE, BTYPE, CTYPE, DTYPE)) {       \
+    return DoMatmul<float, CudaToNativeT<ATYPE>::type,                      \
+                    CudaToNativeT<BTYPE>::type, CudaToNativeT<CTYPE>::type, \
+                    CudaToNativeT<DTYPE>::type>(                            \
+        stream, a_buffer, b_buffer, c_buffer, d_buffer, bias_buffer,        \
+        a_scale_buffer, b_scale_buffer, c_scale_buffer, d_scale_buffer,     \
+        d_amax_buffer, algorithm, scratch_allocator, profile_result);       \
+  }
+
+  // FP8 compatible type combinations (see cuBLASLt documentation):
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E4M3, CUDA_R_16BF, CUDA_R_16BF);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E4M3, CUDA_R_16BF, CUDA_R_8F_E4M3);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E4M3, CUDA_R_16F, CUDA_R_8F_E4M3);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E4M3, CUDA_R_16F, CUDA_R_16F);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E4M3, CUDA_R_32F, CUDA_R_32F);
+
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E5M2, CUDA_R_16BF, CUDA_R_16BF);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E5M2, CUDA_R_16BF, CUDA_R_8F_E4M3);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E5M2, CUDA_R_16BF, CUDA_R_8F_E5M2);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E5M2, CUDA_R_16F, CUDA_R_8F_E4M3);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E5M2, CUDA_R_16F, CUDA_R_8F_E5M2);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E5M2, CUDA_R_16F, CUDA_R_16F);
+  TYPED_MATMUL_F8(CUDA_R_8F_E4M3, CUDA_R_8F_E5M2, CUDA_R_32F, CUDA_R_32F);
+
+  TYPED_MATMUL_F8(CUDA_R_8F_E5M2, CUDA_R_8F_E4M3, CUDA_R_16BF, CUDA_R_16BF);
+  TYPED_MATMUL_F8(CUDA_R_8F_E5M2, CUDA_R_8F_E4M3, CUDA_R_16BF, CUDA_R_8F_E4M3);
+  TYPED_MATMUL_F8(CUDA_R_8F_E5M2, CUDA_R_8F_E4M3, CUDA_R_16BF, CUDA_R_8F_E5M2);
+  TYPED_MATMUL_F8(CUDA_R_8F_E5M2, CUDA_R_8F_E4M3, CUDA_R_16F, CUDA_R_8F_E4M3);
+  TYPED_MATMUL_F8(CUDA_R_8F_E5M2, CUDA_R_8F_E4M3, CUDA_R_16F, CUDA_R_8F_E5M2);
+  TYPED_MATMUL_F8(CUDA_R_8F_E5M2, CUDA_R_8F_E4M3, CUDA_R_16F, CUDA_R_16F);
+  TYPED_MATMUL_F8(CUDA_R_8F_E5M2, CUDA_R_8F_E4M3, CUDA_R_32F, CUDA_R_32F);
+
+#undef TYPED_MATMUL_F8
+
+  // Other data types:
+  if (operand_types ==
+      std::make_tuple(CUDA_R_16F, CUDA_R_16F, CUDA_R_16F, CUDA_R_16F)) {
```

Comment:
Can these cases also be handled by the macro?

Reply:
Done.